File: output/0322.零钱兑换/0322-零钱兑换.go
// 背包问题
func coinChange1(coins []int, amount int) int {
    dp := make([]int, amount+1)
	dp[0] = 0
	n := len(coins)

    for i:=1; i<=amount; i++{
        res := math.MaxInt64
        for j:=0; j<n; j++{
            if i == coins[j]{
                dp[i] = 1
                break
            }else if i > coins[j]{
                if dp[i - coins[j]] > 0{
                    res = min(res, dp[i - coins[j]])
                }
            }
            if res != math.MaxInt64{
                dp[i] = res + 1
            }else{
                dp[i] = -1
            }
        }
    }
    return dp[amount]
}


func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	dp[0] = 0
	n := len(coins)

	for i := 1; i <= amount; i++ {
		dp[i] = amount + 1
		for j := 0; j < n; j++ {
			if i >= coins[j] {
				dp[i] = min(dp[i], dp[i-coins[j]]+1)
			}
		}
	}

	if amount < dp[amount] {
		return -1
	}

	return dp[amount]
}

File: output/0322.零钱兑换/0322-零钱兑换.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if not amount:
            return 0
        dp = [0] *(amount + 1)
        for c in coins:
            if c < amount + 1:
                dp[c] = 1
        for i in range(amount+1):
            now = [dp[i]] if dp[i] else []
            for c in coins:
                if i - c >= 0 and dp[i-c]:
                    now.append(dp[i-c] + 1)
            if now:
                dp[i] = min(now)

        return dp[-1] if dp[-1] != 0 else -1